File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        HashMap<Character, Integer> freq = new HashMap<>();
        for(int i=0;i<s.length();i++){
            char key = s.charAt(i);
            if (freq.containsKey(key)) { 
                freq.put(key, freq.get(key) + 1);
            } else {
                freq.put(key, 1);
            }

        }
        
        ArrayList<Map.Entry<Character, Integer>> list = new ArrayList<>(freq.entrySet());

        // Step 3: Sort by frequency in descending order
        Collections.sort(list, new Comparator<Map.Entry<Character, Integer>>() {
            public int compare(Map.Entry<Character, Integer> a, Map.Entry<Character, Integer> b) {
                return b.getValue() - a.getValue(); // Descending
            }
        });

        // Step 4: Build the output string
        StringBuilder result = new StringBuilder();
        for (Map.Entry<Character, Integer> entry : list) {
            char ch = entry.getKey();
            int count = entry.getValue();
            for (int i = 0; i < count; i++) {
                result.append(ch);
            }
        }

        return result.toString();
    }
}

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.go
func frequencySort(s string) string {
    freq:=make(map[rune]int);

    for _,c:=range s{
        freq[c]++;
    }

    n:=len(s)

    bucket:=make([][]rune,n+1);
    for ch,count:=range freq{
        bucket[count]=append(bucket[count],ch)
    }

    result:=make([]rune,0,n)
    for i:=n;i>0;i--{
        for _,c:=range bucket[i]{
            for j:=0;j<i;j++{
                result=append(result,c)
            }
        }
    }

    return string(result)
}